File: DevEnvironment/DataStructures/src/main/java/datastructures/bitarray/core/Bit32ArrayIterator.java
package datastructures.bitarray.core;

import base.core.Conditions;
import base.interfaces.IIterator;
import datastructures.bitarray.interfaces.IBit32Array;

/**
 * The Bit32ArrayIterator class implements an iterator of a bit 32 array.
 */
public final class Bit32ArrayIterator implements IIterator<Boolean> {
    private final IBit32Array bitArray;
    private int currIndex;

    /**
     * Creates an iterator of a bit 32 array.
     */
    public static IIterator<Boolean> make(IBit32Array bitArray) {
        return new Bit32ArrayIterator(bitArray);
    }

    /**
     * The BitArrayIterator constructor.
     */
    private Bit32ArrayIterator(IBit32Array bitArray) {
        Conditions.validateNotNull(
            bitArray,
            "The bit 32 array.");

        this.bitArray = bitArray;

        this.reset();
    }

    /**
     * Checks whether there is a next element.
     */
    @Override
    public boolean hasNext() {
        return this.currIndex < this.bitArray.size();
    }

    /**
     * Gets the next element.
     */
    @Override
    public Boolean next() {
        assert(this.hasNext());

        boolean currElement = this.bitArray.isOn(this.currIndex);
        ++this.currIndex;

        return currElement;
    }

    /**
     * Resets the iterator.
     */
    @Override
    public void reset() {
        this.currIndex = 0;
    }
}

File: SystemCore/DataStructures/bitarray/core/Bit32ArrayIterator.cpp
#include "Bit32ArrayIterator.h"
#include "IBit32Array.h"
#include <assert.h>

using namespace datastructures::bitarray;

/**
 * Creates an iterator of a bit 32 array.
 */
base::IIteratorSharedPtr<bool> Bit32ArrayIterator::Make(const IBit32Array& bitArray)
{
    return std::make_shared<Bit32ArrayIterator>(bitArray);
}

/**
 * The Bit32ArrayIterator constructor.
 */
Bit32ArrayIterator::Bit32ArrayIterator(const IBit32Array& bitArray) :
    m_bitArray(bitArray)
{
    Reset();
}

/**
 * The Bit32ArrayIterator destructor.
 */
Bit32ArrayIterator::~Bit32ArrayIterator()
{
}

/**
 * Checks whether there is a next element.
 */
bool Bit32ArrayIterator::HasNext() const
{
    return m_currIndex < m_bitArray.Size();
}

/**
 * Gets the next element.
 */
bool Bit32ArrayIterator::Next()
{
    assert(HasNext());

    bool currElement = m_bitArray.IsOn(m_currIndex);
    ++m_currIndex;

    return currElement;
}

/**
 * Resets the iterator.
 */
void Bit32ArrayIterator::Reset()
{
    m_currIndex = 0;
}
